rslib/src/import_export/text/import.rs

```
@@ -0,0 +1,216 @@
+}
+
+impl<'a> Context<'a> {
+    fn new(data: &ForeignData, col: &'a mut Collection) -> Result<Self> {
+        let usn = col.usn()?;
+        let normalize_notes = col.get_config_bool(BoolKey::NormalizeNoteText);
+        let mut notetypes = HashMap::new();
+        notetypes.insert(
+            String::new(),
+            col.notetype_for_string(&data.default_notetype)?,
+        );
+        let mut deck_ids = HashMap::new();
+        deck_ids.insert(String::new(), col.deck_id_for_string(&data.default_deck)?);
+        Ok(Self {
+            col,
+            usn,
+            normalize_notes,
+            notetypes,
+            deck_ids,
+        })
+    }
+
+    fn import_foreign_notetypes(&mut self, notetypes: Vec<ForeignNotetype>) -> Result<()> {
+        for foreign in notetypes {
+            let mut notetype = foreign.into_native();
+            notetype.usn = self.usn;
+            self.col
+                .add_notetype_inner(&mut notetype, self.usn, false)?;
+        }
+        Ok(())
+    }
+    fn notetype_for_note(&mut self, note: &ForeignNote) -> Result<Option<Arc<Notetype>>> {
+        Ok(if let Some(nt) = self.notetypes.get(&note.notetype) {
+            nt.clone()
+        } else {
+            let nt = self.col.notetype_for_string(&note.notetype)?;
+            self.notetypes.insert(note.notetype.clone(), nt.clone());
+            nt
+        })
+    }
+
+    fn deck_id_for_note(&mut self, note: &ForeignNote) -> Result<Option<DeckId>> {
+        Ok(if let Some(did) = self.deck_ids.get(&note.deck) {
+            *did
+        } else {
+            let did = self.col.deck_id_for_string(&note.deck)?;
+            self.deck_ids.insert(note.deck.clone(), did);
+            did
+        })
+    }
+
+    fn import_foreign_notes(&mut self, notes: Vec<ForeignNote>) -> Result<NoteLog> {
+        let mut log = NoteLog::default();
+        for foreign in notes {
+            if let Some(notetype) = self.notetype_for_note(&foreign)? {
```

Warning or failing might be better than silently ignoring